[Bug] [FailoverCoordinator] Duplicated Fail-overed workflow handling

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

when restart master-cluster, there alive master-server's workflow instance may take-overed by new master instance.

there are some log segments, when  `master-server 172.2.1.20` started, it first take over  `master-server 172.2.1.21`'s workflows(WORKFLOW-A,B,C), then `master-server 172.2.1.21` started, as the logic of `getFailoverWorkflowsForMaster`, there use workflows' start_time to compare with  `master-server 172.2.1.20`'s startup time, which were take-overed by  `master-server 172.2.1.20` just now, and create command to do failover again.

```
====== master-server 172.2.1.20 ======

[WI-0][TI-0] - 2025-07-10 20:30:49.341 INFO  [Master-Server] o.a.d.s.m.e.s.SystemEventBus:[40] - Published SystemEvent: GlobalMasterFailoverEvent{eventTime=Thu Jul 10 20:30:48 GMT+08:00 2025}
[WI-0][TI-0] - 2025-07-10 20:30:49.341 INFO  [Master-Server] o.a.d.s.m.e.s.SystemEventBusFireWorker:[62] - SystemEventBusFireWorker started
[WI-0][TI-0] - 2025-07-10 20:30:49.346 INFO  [Master-Server] o.a.d.s.m.MasterServer:[164] - MasterServer initialized successfully in 1343 ms
[WI-0][TI-0] - 2025-07-10 20:30:49.346 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[73] - Global master failover starting
[WI-0][TI-0] - 2025-07-10 20:30:49.490 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[90] - The master[172.2.1.21:5678] is not alive, do global master failover on it
[WI-0][TI-0] - 2025-07-10 20:30:49.818 INFO  [Curator-TreeCache-0] o.a.d.s.m.c.AbstractClusterSubscribeListener:[41] - Server WorkerServerMetadata(workerGroup=default, workerWeight=100.0, taskThreadPoolUsage=0.0) added
[WI-0][TI-0] - 2025-07-10 20:30:50.117 INFO  [Curator-TreeCache-0] o.a.d.s.m.c.AbstractClusterSubscribeListener:[41] - Server WorkerServerMetadata(workerGroup=default, workerWeight=100.0, taskThreadPoolUsage=0.0) added
[WI-0][TI-0] - 2025-07-10 20:30:50.779 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4826511, name=WORKFLOW-C-3-20250710050000600, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:50.787 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4826537, name=WORKFLOW-B-5-20250710050001109, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:50.793 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4828292, name=WORKFLOW-A-20250710194500099, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:50.795 INFO  [SystemEventBusFireWorker] o.a.d.r.a.RegistryClient:[177] - persist key: /nodes/failover-finish-nodes/172.2.1.21:5678-unknown-unknown, value: 1752150648003
[WI-0][TI-0] - 2025-07-10 20:30:50.803 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[155] - Master[172.2.1.21:5678] failover 3 workflows finished, cost: 1312/ms
[WI-0][TI-0] - 2025-07-10 20:30:50.803 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[99] - Global master failover finished, cost: 1457/ms
[WI-0][TI-0] - 2025-07-10 20:30:50.803 INFO  [SystemEventBusFireWorker] o.a.d.s.m.e.s.SystemEventBusFireWorker:[103] - Fire SystemEvent: GlobalMasterFailoverEvent{eventTime=Thu Jul 10 20:30:48 GMT+08:00 2025} cost: 1459 ms
[WI-0][TI-0] - 2025-07-10 20:30:51.011 INFO  [MasterCommandHandleThreadPool] o.a.d.s.m.e.WorkflowEventBus:[41] - Publish event: WorkflowStartLifecycleEvent{workflow=WORKFLOW-A-20250710194500099}
[WI-0][TI-0] - 2025-07-10 20:30:51.012 INFO  [MasterCommandHandleThreadPool] o.a.d.s.m.e.WorkflowEventBus:[41] - Publish event: WorkflowStartLifecycleEvent{workflow=WORKFLOW-B-5-20250710050001109}
[WI-0][TI-0] - 2025-07-10 20:30:51.012 INFO  [MasterCommandHandleThreadPool] o.a.d.s.m.e.WorkflowEventBus:[41] - Publish event: WorkflowStartLifecycleEvent{workflow=WORKFLOW-C-3-20250710050000600}
[WI-0][TI-0] - 2025-07-10 20:30:51.018 INFO  [MasterCommandHandleThreadPool] o.a.d.s.m.e.c.CommandEngine:[174] - Success bootstrap command {
...


====== master-server 172.2.1.21 ======

[WI-0][TI-0] - 2025-07-10 20:30:53.603 INFO  [Master-Server] o.a.d.s.m.e.s.SystemEventBus:[40] - Published SystemEvent: GlobalMasterFailoverEvent{eventTime=Thu Jul 10 20:30:51 GMT+08:00 2025}
[WI-0][TI-0] - 2025-07-10 20:30:53.603 INFO  [Master-Server] o.a.d.s.m.e.s.SystemEventBusFireWorker:[62] - SystemEventBusFireWorker started
[WI-0][TI-0] - 2025-07-10 20:30:53.609 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[73] - Global master failover starting
[WI-0][TI-0] - 2025-07-10 20:30:53.612 INFO  [Master-Server] o.a.d.s.m.MasterServer:[164] - MasterServer initialized successfully in 1670 ms
[WI-0][TI-0] - 2025-07-10 20:30:53.720 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[82] - The master[MasterServerMetadata(super=BaseServerMetadata(processId=9652, serverStartupTime=1752150648003, address=172.2.1.20:5678, cpuUsage=0.02604212364461751, memoryUsage=0.33144767254423296, serverStatus=NORMAL))] is alive, do global master failover on it
[WI-0][TI-0] - 2025-07-10 20:30:53.927 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4826511, name=WORKFLOW-C-3-20250710050000600, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:53.933 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4826537, name=WORKFLOW-B-5-20250710050001109, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:53.940 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.WorkflowFailover:[64] - Success failover workflowInstance: [id=4828292, name=WORKFLOW-A-20250710194500099, state=RUNNING_EXECUTION]
[WI-0][TI-0] - 2025-07-10 20:30:53.942 INFO  [SystemEventBusFireWorker] o.a.d.r.a.RegistryClient:[177] - persist key: /nodes/failover-finish-nodes/172.2.1.20:5678-unknown-unknown, value: 1752150648003
[WI-0][TI-0] - 2025-07-10 20:30:53.953 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[155] - Master[172.2.1.20:5678] failover 3 workflows finished, cost: 231/ms
[WI-0][TI-0] - 2025-07-10 20:30:53.953 INFO  [SystemEventBusFireWorker] o.a.d.s.m.f.FailoverCoordinator:[82] - The master[MasterServerMetadata(super=BaseServerMetadata(processId=28662, serverStartupTime=1752150651942, address=172.2.1.21:5678, cpuUsage=0.025943546541755118, memoryUsage=0.3181573051280973, serverStatus=NORMAL))] is alive, do global master failover on it
```

so i think there need a host-timestamp (host address + host startup time) to mark the workflow-instance(e.g. set to `workflowinstance.host`), to decide whether the workflow should be takeover, This approach may be more stable and simple.

### What you expected to happen

-

### How to reproduce

-

### Anything else

also, i found a new issue here, when workflow take-overed by another master, the unreported task executor event will stack in executor, i will create a new issue. 

### Version

dev

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

> We do update the workflow's host, there might exist concurrent issue, it's better to get the workflow's host from executor, once the event ready to report, then get the host from executor.
> 
> ```
> public boolean reassignWorkflowInstanceHost(final TaskExecutorReassignMasterRequest taskExecutorReassignMasterRequest) {
>         final int taskInstanceId = taskExecutorReassignMasterRequest.getTaskInstanceId();
>         final String workflowHost = taskExecutorReassignMasterRequest.getWorkflowHost();
>         // todo: Is this reassign can make sure there is no concurrent problem?
>         physicalTaskExecutorRepository.get(taskInstanceId).ifPresent(
>                 taskExecutor -> taskExecutor.getTaskExecutionContext().setWorkflowInstanceHost(workflowHost));
>         return physicalTaskExecutorEventReporter.reassignWorkflowInstanceHost(taskInstanceId, workflowHost);
>     }
> ```

Oh i found why! it's caused by this issue, there are the other logs:
```
[WI-0][TI-0] - 2025-07-10 20:30:54.101 ERROR [MasterCommandHandleThreadPool] o.a.d.s.m.e.c.CommandEngine:[186] - Failed bootstrap command {
  "id" : 4889016,
  "commandType" : "RECOVER_TOLERANCE_FAULT_PROCESS",
  "workflowDefinitionCode" : 15081302155680,
  "workflowDefinitionVersion" : 20,
  "workflowInstanceId" : 4828292,
  "commandParam" : "{\"commandType\":\"RECOVER_TOLERANCE_FAULT\",\"subWorkflowInstance\":false,\"startNodes\":null,\"commandParams\":null,\"timeZone\":null,\"workflowExecutionStatus\":\"RUNNING_EXECUTION\"}",
  "workflowInstancePriority" : "MEDIUM",
  "executorId" : 0,
  "taskDependType" : "TASK_POST",
  "failureStrategy" : "CONTINUE",
  "warningType" : "NONE",
  "warningGroupId" : null,
  "scheduleTime" : null,
  "startTime" : null,
  "updateTime" : "2025-07-10 20:30:53",
  "workerGroup" : null,
  "tenantCode" : "default",
  "environmentCode" : -1,
  "dryRun" : 0
} 
java.util.concurrent.CompletionException: java.lang.IllegalStateException: WorkflowExecuteRunnable(4828292/WORKFLOW-A-20250710194500099 already registered at WorkflowEventBusFireWorker
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:673)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1609)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: WorkflowExecuteRunnable(4828292/WORKFLOW-A-20250710194500099 already registered at WorkflowEventBusFireWorker
        at com.google.common.base.Preconditions.checkState(Preconditions.java:821)
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusFireWorker.registerWorkflowEventBus(WorkflowEventBusFireWorker.java:63)
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusCoordinator.registerWorkflowEventBus(WorkflowEventBusCoordinator.java:50)
        at org.apache.dolphinscheduler.server.master.engine.command.CommandEngine.bootstrapWorkflowExecutionRunnable(CommandEngine.java:167)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670)
        ... 6 common frames omitted
```

after the master-2.1.21 started, it published the failover command too, coincidentally, master 2.1.20 captured this command, after called `bootstrapCommand` in `CommandEngine`, it failed on `bootstrapWorkflowExecutionRunnable`.

the new `workflowExecutionRunnable` is already put into `workflowRepository`, but failed in `workflowEventBusCoordinator.registerWorkflowEventBus`, so there is no thread to handle the new `workflowExecutionRunnable`'s event bus, so all the events got stuck in the executor's channel.

so it's not the issue of task executor's host reassign. @ruanwenjun

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/MasterServer.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/MasterServer.java
@@ -117,7 +117,6 @@ public class MasterServer implements IStoppable {
     @PostConstruct
     public void initialized() {
         ServerLifeCycleManager.toRunning();
-        final long startupTime = System.currentTimeMillis();
 
         // init rpc server
         this.masterRPCServer.start();
@@ -140,7 +139,8 @@ public class MasterServer implements IStoppable {
 
         this.schedulerApi.start();
 
-        this.systemEventBus.publish(GlobalMasterFailoverEvent.of(new Date(startupTime)));
+        this.systemEventBus
+                .publish(GlobalMasterFailoverEvent.of(new Date(ServerLifeCycleManager.getServerStartupTime())));
         this.systemEventBusFireWorker.start();
 
         MasterServerMetrics.registerMasterCpuUsageGauge(() -> {
@@ -161,7 +161,8 @@ public class MasterServer implements IStoppable {
                 close("MasterServer shutdownHook");
             }
         }));
-        log.info("MasterServer initialized successfully in {} ms", System.currentTimeMillis() - startupTime);
+        log.info("MasterServer initialized successfully in {} ms",
+                System.currentTimeMillis() - ServerLifeCycleManager.getServerStartupTime());
     }
 
     @PreDestroy

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/command/handler/WorkflowFailoverCommandHandler.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/command/handler/WorkflowFailoverCommandHandler.java
@@ -33,6 +33,7 @@ import org.apache.dolphinscheduler.server.master.engine.task.runnable.TaskExecut
 import org.apache.dolphinscheduler.server.master.engine.task.runnable.TaskExecutionRunnableBuilder;
 import org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteContext.WorkflowExecuteContextBuilder;
 
+import java.util.Date;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.BiConsumer;
@@ -89,6 +90,7 @@ public class WorkflowFailoverCommandHandler extends AbstractCommandHandler {
             throw new IllegalArgumentException(
                     "The WorkflowFailoverCommandParam: " + command.getCommandParam() + " is invalid");
         }
+        workflowInstance.setRestartTime(new Date());
         workflowInstance.setState(workflowFailoverCommandParam.getWorkflowExecutionStatus());
         workflowInstance.setHost(masterConfig.getMasterAddress());
         workflowInstanceDao.updateById(workflowInstance);

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/failover/FailoverCoordinator.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/failover/FailoverCoordinator.java
@@ -83,7 +83,7 @@ public class FailoverCoordinator implements IFailoverCoordinator {
                 doMasterFailover(
                         masterAddress,
                         aliveMasterServerMetadata.getServerStartupTime(),
-                        RegistryUtils.getFailoveredNodePathWhichStartupTimeIsUnknown(
+                        RegistryUtils.getGlobalMasterFailoverNodePath(
                                 masterAddress));
             } else {
                 // If the master is not alive, then we use the event time as the failover deadline.
@@ -91,7 +91,7 @@ public class FailoverCoordinator implements IFailoverCoordinator {
                 doMasterFailover(
                         masterAddress,
                         globalMasterFailoverEvent.getEventTime().getTime(),
-                        RegistryUtils.getFailoveredNodePathWhichStartupTimeIsUnknown(masterAddress));
+                        RegistryUtils.getGlobalMasterFailoverNodePath(masterAddress));
             }
         }
 

--- a/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/cases/WorkflowInstanceFailoverTestCase.java
+++ b/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/cases/WorkflowInstanceFailoverTestCase.java
@@ -568,7 +568,7 @@ public class WorkflowInstanceFailoverTestCase extends AbstractMasterIntegrationT
 
         systemEventBus.publish(GlobalMasterFailoverEvent.of(new Date()));
 
-        final String masterFailoverNodePath = RegistryUtils.getFailoveredNodePathWhichStartupTimeIsUnknown(
+        final String masterFailoverNodePath = RegistryUtils.getGlobalMasterFailoverNodePath(
                 "127.0.0.1:15678");
         // wait failover process
         await()

--- a/dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/utils/RegistryUtils.java
+++ b/dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/utils/RegistryUtils.java
@@ -29,7 +29,7 @@ public class RegistryUtils {
         return RegistryNodeType.MASTER_FAILOVER_LOCK.getRegistryPath() + "/" + masterAddress;
     }
 
-    public static String getFailoveredNodePathWhichStartupTimeIsUnknown(final String serverAddress) {
+    public static String getGlobalMasterFailoverNodePath(final String serverAddress) {
         return RegistryNodeType.FAILOVER_FINISH_NODES.getRegistryPath() + "/" + serverAddress + "-" + "unknown" + "-"
                 + "unknown";
     }

